File: mpi_adv_examples/pi_monte.f90
! Example Name  : pi_monte.f90
! Compile   : $ mpif90 -g -o pi_monte.x -Wall pi_monte.f90
! Run       : $ mpirun -np 4 -hostfile hosts pi_monte.x

PROGRAM pi_monte
    IMPLICIT NONE
    INCLUDE 'mpif.h'
    INTEGER nRank, nProcs, iproc, iErr
    INTEGER :: ROOT = 0, scope = 100000000
    INTEGER :: i, nMyCount = 0, nCount = 0
    REAL :: x, y, z, pi, z1

    INTEGER status(MPI_STATUS_SIZE)

    CALL MPI_INIT(iErr)
    CALL MPI_COMM_RANK(MPI_COMM_WORLD, nRank, iErr)
    CALL MPI_COMM_SIZE(MPI_COMM_WORLD, nProcs, iErr)

    ! Get Random #               
    CALL INIT_RANDOM_SEED(nRank+1)
        
    DO i=0, scope-1
        CALL RANDOM_NUMBER(x)
        CALL RANDOM_NUMBER(y)
        z = x*x + y*y
        z1 = SQRT(z)

        IF (z1 <= 1) nMyCount = nMyCount + 1
    END DO

    IF (nRank == ROOT) THEN
        nCount = nMyCount

        DO iproc=1, nProcs-1
            CALL MPI_RECV(nMyCount, 1, MPI_INTEGER, iproc, 55, MPI_COMM_WORLD, &
                            status, iErr)
            nCount = nCount + nMyCount
        END DO
    
        pi = 4 * REAL(nCount) / (scope * nProcs)

        WRITE (*, '(A, I2, A, I10, A)') 'Processor=', nRank, ' sending results = ', &
                nMyCount, ' to ROOT process'
        WRITE (*, '(A, I2, A, F15.10)') '# of trial is ', nProcs, ' estimate of PI is ', pi

    ELSE
        CALL MPI_SEND(nMyCount, 1, MPI_INTEGER, ROOT, 55, MPI_COMM_WORLD, iErr)
        WRITE (*, '(A, I2, A, I10)') 'Processor=', nRank, ' sending results = ', nMyCount
    END IF
    CALL MPI_FINALIZE(iErr)

    CONTAINS

    SUBROUTINE INIT_RANDOM_SEED(rank)
        IMPLICIT NONE
        INTEGER rank
        INTEGER :: i, n, clock
        INTEGER, DIMENSION(:), ALLOCATABLE :: seed
                
        CALL RANDOM_SEED(size = n)
        ALLOCATE(seed(n))
                
        CALL SYSTEM_CLOCK(COUNT=clock)
    
        seed = clock + 37 * (/ (i - 1, i = 1, n) /)
        seed = seed * rank * rank
    
        CALL RANDOM_SEED(PUT = seed)
                
        DEALLOCATE(seed)
    END SUBROUTINE
        
END
                

File: mpi_adv_examples/pi_monte.c
/*
	Example Name	: pi_monte.c
	Compile	: $ mpicc -g -o pi_monte -Wall pi_monte.c
	Run		: $ mpirun -np 4 -hostfile hosts pi_monte
*/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <mpi.h>

#define	SCOPE	100000000

/*
** PI (3.14) Monte Carlo Method
*/
int main(int argc, char *argv[])
{
	int nProcs, nRank, proc, ROOT = 0;
	MPI_Status status;
	int nTag = 55;

	int i, nCount = 0, nMyCount = 0;
	double x, y, z, pi, z1;
	MPI_Init(&argc, &argv);
	MPI_Comm_size(MPI_COMM_WORLD, &nProcs);
	MPI_Comm_rank(MPI_COMM_WORLD, &nRank);

	srand(time(NULL)*(nRank+1));

	for (i = 0; i < SCOPE; i++) {
		x = (double)rand() / (double)(RAND_MAX);
		y = (double)rand() / (double)(RAND_MAX);

		z = x*x + y*y;
		z1 = sqrt(z);

		if (z1 <= 1)	nMyCount++;
	}

	if (nRank == ROOT) {
		nCount = nMyCount;

		for (proc=1; proc<nProcs; proc++) {
			MPI_Recv(&nMyCount, 1, MPI_REAL, proc, nTag, MPI_COMM_WORLD,
                             &status);
			nCount += nMyCount;
		}

		pi = (double)4*nCount/(SCOPE * nProcs);
		printf("Processor %d sending results = %d to ROOT processor\n", nRank,
                 nMyCount);
		printf("\n # of trials (cpu#: %d, time#: %d) = %d, estimate of pi is %f\n",
				nProcs, SCOPE, SCOPE*nProcs, pi);

	}
	else {
		printf("Processor %d sending results = %d to ROOT processor\n", nRank,
                 nMyCount);
		MPI_Send(&nMyCount, 1, MPI_REAL, ROOT, nTag, MPI_COMM_WORLD);
	}

	printf("\n");

	MPI_Finalize();

	return 0;
}
